web/vtadmin/src/api/http.test.ts

```diff
@@ -0,0 +1,161 @@
+// malformed payloads, and timeouts.
+//
+// The big downside to mocking or "faking" APIs like vtadmin is that
+// we end up re-implementing some (or all) of vtadmin-api in our test environment.
+// It is, unfortunately, impossible to completely avoid this kind of duplication
+// unless we solely use e2e tests (which have their own trade-offs).
+//
+// That said, our use of protobufjs to validate and strongly type HTTP responses
+// means our fake is more robust than it would be otherwise. Since we are using
+// the exact same protos in our fake as in our real vtadmin-api server, we're guaranteed
+// to have type parity.
+//
+process.env.REACT_APP_VTADMIN_API_ADDRESS = '';
+const server = setupServer();
+
+const mockServerJson = (endpoint: string, json: object) => {
+    server.use(rest.get(endpoint, (req, res, ctx) => res(ctx.json(json))));
+};
+
+// Enable API mocking before tests.
+beforeAll(() => server.listen());
+
+// Reset any runtime request handlers we may add during the tests.
+afterEach(() => server.resetHandlers());
+
+// Disable API mocking after the tests are done.
+afterAll(() => server.close());
+
+describe('api/http', () => {
+    describe('vtfetch', () => {
+        it('parses and returns JSON, given an HttpOkResponse response', async () => {
+            const endpoint = `/api/tablets`;
+            const response = { ok: true, result: null };
+            mockServerJson(endpoint, response);
+
+            const result = await api.vtfetch(endpoint);
+            expect(result).toEqual(response);
+        });
+
+        it('parses and returns JSON, given an HttpErrorResponse response', async () => {
+            const endpoint = `/api/tablets`;
+            const response = { ok: false };
+            mockServerJson(endpoint, response);
+
+            const result = await api.vtfetch(endpoint);
+            expect(result).toEqual(response);
+        });
+
+        it('throws an error on malformed JSON', async () => {
+            const endpoint = `/api/tablets`;
+            server.use(rest.get(endpoint, (req, res, ctx) => res(ctx.body('this is fine'))));
+
+            expect.assertions(2);
+
+            try {
+                await api.vtfetch(endpoint);
+            } catch (e) {
+                /* eslint-disable jest/no-conditional-expect */
+                expect(e.name).toEqual('SyntaxError');
+                expect(e.message.startsWith('Unexpected token')).toBeTruthy();
+                /* eslint-enable jest/no-conditional-expect */
+            }
+        });
+
+        it('throws an error on malformed response envelopes', async () => {
+            const endpoint = `/api/tablets`;
+            mockServerJson(endpoint, { foo: 'bar' });
+
+            expect.assertions(1);
+
+            try {
+                await api.vtfetch(endpoint);
+            } catch (e) {
+                /* eslint-disable jest/no-conditional-expect */
+                expect(e.message).toEqual('invalid http envelope');
```

Comment:
One of the potential downsides to doing exact equality checks on error messages means that if we ever add additional context to errors, then we can potentially break a bunch of tests for an unrelated change. Do we want to force that, or does it make sense to `expect().contains()`?

Reply:
Hm, probably the best way is to [extend `Error` with custom subclasses](https://javascript.info/custom-errors). Unfortunately, this is pretty cumbersome to do in JS/TS (especially compared to go 🥺) BUT we can then do assertions like `expect(e.name).toEqual('InvalidHttpEnvelope')` or whatever, which also gives us more flexibility with `e.message`.

I considered this before and didn't do it because I'm uh... well, it's a little bit of work 😬 but it's super worth it, so thanks for pointing this out!